osc: improve look of seekranges

medhefgo opened the pull request:
This looks much better imho. In case of the bar style we won't see seek ranges in the past, but that is of little use. If desired I can also make this change into a seekbarstyle option for users to chose from.

See the screenshot for comparison. Top 3 is before and bottom 3 are after. Note that slider style wasn't changed at all since that looks fine (and doesn't really work with the way the seek ranges are rendered now).
![seekranges](https://user-images.githubusercontent.com/7892635/35038395-ffe25a18-fb7a-11e7-914d-f332a679c46b.jpeg)

wm4 commented:
It's nice that someone is looking into this.

The first new example (4/6) doesn't look so ideal, because the left part of the bar completely hides the seek range.

medhefgo commented:
Yes, that's true. But I'd say that people hardly care about seek ranges that lie in the past.

wm4 commented:
Well, not sure if I agree with that. If you seek and care about the seek ranges, you want to use them to decide whether you can seek there quickly. And seeking backwards is a pretty common thing.

medhefgo commented:
Here's a wip how it could look by just adding an alpha channel to the bar. A proper version ought to render the bar opaque but leave out holes for the seek ranges. Doable, but is it worth the effort? I kinda dislike how it looks. Also, it makes the current playback position less clear imho.
![wip](https://user-images.githubusercontent.com/7892635/35040687-6892c348-fb82-11e7-82b0-856080a69763.jpeg)

wm4 commented:
Looks a bit confusing, but I'm not a UI person.

wiiaboo commented:
For example 1|4 you could also try moving the cache lines to either bottom or top and make them thicker.

medhefgo commented:
Here's another go at it. The second one looks okay-ish I guess. I still prefer my original choice.
![wip](https://user-images.githubusercontent.com/7892635/35041678-e0548346-fb85-11e7-916b-4c56bf3da08f.jpeg)

wiiaboo commented:
Leave the same contrast as before and over the white bar. And use the second one, yes.

wm4 commented:
We could also just change the default slider style, maybe?

wiiaboo commented:
For that I'd really prefer going with a poll of some sort and leave it like a week or two up. Wouldn't want a repeat of the layout change.

wm4 commented:
Github can effectively do polls with "reactions" to github posts. Should we try that? (Like making a new issue, instructing people to use specific reactions for voting for a thing, and then spread the issue link.)

medhefgo commented:
Urgh. The other bar styles are just too ugly to make them default.

Here's one with it drawn over the white bar. Doesn't exactly look as I'd hoped.
![wip](https://user-images.githubusercontent.com/7892635/35046651-e8b55102-fb97-11e7-8f2a-8ab8d43ad176.jpeg)

wiiaboo commented:
> Urgh. The other bar styles are just too ugly to make them default.

That's the thing. They're all so subjective. Since knob was added that's the one I've been using forever.

> Here's one with it drawn over the white bar.

Doesn't look worse than the other one with absolutely no contrast.

paulguy commented:
I like the seek ranges below the bar, and maybe make the bar fill the whole width of the channel if it's just a local file or something.  Could even make the seek ranges a different color or something if contrast is too low.  The ranges over the bar looks a bit ew in my opinion but yeah I'm just one user so i'll vote when the poll comes out.

kevmitch commented:
Don't know how helpful this is since my tastes are probably weird, but . . .

I like the asthetics of the round `knob`, but not so much function (or rather lack there of) of the rail it's on because it seems redundant except for indicating the range of the centre of the knob. I find the padding regions on the ends a little weird (though I guess this is supposed to be a metaphor for a physical thing). Could the light area be the range of possible centre positions instead and the knob would straddle the dark and light at the start and end?

The `knob` combined with the way the seek ranges are displayed with `slider` would be ideal for me because each visual element would communicate something non-trivial. The past / future asymmetry of the `box` and the extraneous rail of `knob` complicate the seek ranges visualization for me.

quilloss commented:
How about reducing the thickness of `bar` and center it with the seek range in the background. That way you get some symmetry in the design.

medhefgo commented:
@quilloss I don't quite get what you mean.

paulguy commented:
I think they mean have the seekbar narrower than the bar showing seek ranges and having the seek ranges show through on the top and bottom of the bar, rather than just showing through the bottom.

medhefgo commented:
Like this?
![wip](https://user-images.githubusercontent.com/7892635/35142706-66826c38-fcff-11e7-98ff-59c6228d21af.jpeg)

shinchiro commented:
Or maybe leave it at default for `bar` style

wiiaboo commented:
@medhefgo how does that look with the floating text?

medhefgo commented:
You mean time code tooltip? First 2 variants are fine. the last one doesn't look ideal imho.
![wip](https://user-images.githubusercontent.com/7892635/35158022-17b80e28-fd36-11e7-83a9-bf0c53e15738.jpeg)

wm4 commented:
Does @ChrisK2 have any opinions?

medhefgo commented:
I force pushed a new version. I made this into a option which defaults to bar.

Now here's my reasoning why bar ought to be the default:
- It's what pretty much every player in the planet uses to signal cached parts
- It's easy to grasp and easy to read. The line style quickly becomes too small to see when the seekbar vertical size becomes too small (small window size or box layout with medium window size).
- Other styles suffer pretty much from being ugly, very hard to grasp from the start or fail to be legible under small vertical sizes.

Note that I decided to hide the "Cache:" part in the osc if visual seekranges are enabled, since I find showing both to be redundant.

From top to bottom: "bar - bar", "bar - line", "slider - bar", "slider - line"
![wip](https://user-images.githubusercontent.com/7892635/35160781-7cb33b24-fd3e-11e7-8211-7aae138b548b.jpeg)

kevmitch commented:
I think these are in the right direction. I think slider-bar is nearly perfect for me. It probably still has that padding at the edges. Could the start and end be shaped to fit the slider when it is there?
Is there a "knob  - bar" equivalent"?

Is the "bar - bar" that you're advocating for the default still not showing past seek ranges? I agree with @wm4 that showing only future seek ranges is not a good idea. I know I generally seek backward more often than I seek forward to review something that I missed or that was important. It's also nice to know how much history of a live stream you still have.

Not sure about removing the actual Cache numbers. Visual information is best complimented by quantitative information if it can fit, and in this case, there's a big space available for it.

Hrxn commented:
> Note that I decided to hide the "Cache:" part in the osc if visual seekranges are enabled, since I find showing both to be redundant.

Absolutely agree here..

wm4 commented:
Technically it's not redundant. The seek ranges show where you can seek to within the cache, while the cache amount includes data that can not be used for seeking yet.

ChrisK2 commented:
> Does @ChrisK2 have any opinions?

Yes, everyone please think of the chapter markers at the top and bottom when you are putting seek ranges there!

Hrxn commented:
Okay, technically it's not redundant, right. But I still think, arguing from the end-user perspective, that this cache amount display is rather confusing. Constantly flickering numbers on the OSC. In my opinion this is technical information, similar to the functionality of `stats`, that should be displayed optionally, or for debug purposes etc.
I don't know why this would be necessary otherwise. Feel free to disagree, of course.

wiiaboo commented:
They shouldn't be that flickery anymore. It only shows demuxer-cache-duration now, which is mostly relevant for livestreams. The refresh rate is also limited depending on how high the number is.

schmied commented:
Thanks for working on this! I already thought about this GUI problem with the new demux cache too. I agree that this has a high bikeshed potential, but for myself I came to this solution: similar to the top OSC of your last picture (bar-bar), but the cache bar with hight transparency and the seek bar with low transparency, so that both add up. So you can see the cached ranges in the past, that are a bit darker than the bright seek-bar. Thanks again!

wm4 commented:
Sadly this got lost in bikeshed. To pick this up again...

Looking at https://github.com/mpv-player/mpv/pull/5411#issuecomment-359019126 I think the minimum requirement is that not only forward cached parts can be distinguished, but also backward cache. bar-bar doesn't seem to satisfy this. Personally I'd be fine with any other style.

medhefgo commented:
Okay, I finally decided to give this another try and refined the look quite a bit.

The the knob and slider are now separated. Instead the slider is now a seekrange option. I also went and made the surrounding box and seekranges visually mimic the shape of the diamond or knob. This includes the shape of slimbox layout and the seekbar border in box layout, I just didn't make a screenshot for that.

Any suggestions for fine tuning the looks are welcome, but please don't ask for any more options and styles or I am gonna lose my sanity.

Now before I update the pull request we should decide on the color of our bike shed. So spread the word: https://www.strawpoll.me/15708883

![samples](https://user-images.githubusercontent.com/7892635/40118123-5172e46e-5919-11e8-98f4-2b5e28ef991e.png)

wm4 commented:
So just technically, 1. is out because it doesn't show the seek ranges in already played parts. 3-8 have the problem that the knob covers the seek range (visually they look ok though). So only 2. would be a suitable default IMO.

schmied commented:
Just to make clear what I meant in my previous post:

https://i.imgur.com/dXeMjUA.png

medhefgo commented:
@wm4 Strictly speaking, the diamond's vertical center is easy to tell and is the part that signals the time position.

@schmied I already did that. See my earlier comments. It's a visually too hard to tell what part is for what, no matter how you fiddle with the alpha.

wm4 commented:
Maybe what I said before was only in context of hiding the seek ranges before the current playback position, but in general no part of the seek ranges should be hidden. And that includes parts hidden by the knob.

medhefgo commented:
Note that the handle size can be reduced. Then it wouldn't cover that much of the seek range anymore.

medhefgo commented:
I've updated the pull request. bar+line is the default style.

jeeb commented:
Thanks for updaitng the PR, will try to have it checked out.

leavittx commented:
Slight offtopic: where could I find any info on seekranges themselves? I want to know whether it's possible to control the amount of data which can be inside the seekrange (let's say I want to be always to quickly seek backwards), and also how do I get their bounds when using libmpv interface? Thanks!

Akemi commented:
i meant to look at this a lot sooner and i finally got around to do it. i made some screenshots of the latest version. overall it's a pretty good improvement, though i have some gripes with it.

1. bar-bar is kinda useless for backwards seek ranges. the only thing i see, that can be done, is to do it similar to bar-line and add a margin at the bottom.
2. a rather small problem with the knob/diamond style is that one can't see the seek ranges that are underneath the knob/diamond. for the line/slider one could overlay the range so they are visible. with the bar that is a bit harder and would look messy.
3. for bar-line i am not a big fan that the bar for the current position is underneath the seekrage and shows through in the none seekable gaps (it changes the colour of the gaps). i personally would rather have that bar smaller.

the seekbar style bar seem to be the most problematic. though if not default it's probably not worth trying to 'fix'. bar-line should probably be the default, because everything is visible there. maybe some of the shortcomings could be mentioned in the manual.

otherwise this looks good. i tested everything for it's functionality and couldn't find any regressions.

**bar bar**
![bar bar](https://user-images.githubusercontent.com/680386/47966850-2d349f00-e057-11e8-9997-42bfa1d68777.png)
**bar line**
![bar line](https://user-images.githubusercontent.com/680386/47966851-2d349f00-e057-11e8-823d-860cb59deb41.png)
**diamond bar**
![diamond bar](https://user-images.githubusercontent.com/680386/47966852-2dcd3580-e057-11e8-9404-0db32a2df721.png)
**diamond line**
![diamond line](https://user-images.githubusercontent.com/680386/47966853-2dcd3580-e057-11e8-89e8-cd4d04fce5e8.png)
**diamond slider**
![diamond slider](https://user-images.githubusercontent.com/680386/47966854-2dcd3580-e057-11e8-9fd2-2e873c064f0e.png)
**knob bar**
![knob bar](https://user-images.githubusercontent.com/680386/47966856-2dcd3580-e057-11e8-8d45-a83978e46503.png)
**knob line**
![knob line](https://user-images.githubusercontent.com/680386/47966857-2dcd3580-e057-11e8-9f87-9da85b35c271.png)
**knob slider**
![knob slider](https://user-images.githubusercontent.com/680386/47966858-2dcd3580-e057-11e8-9756-0e845498bfb5.png)

**bar bar**
![bar bar](https://user-images.githubusercontent.com/680386/47966814-02e2e180-e057-11e8-8b34-c2862afe576d.png)
**bar line**
![bar line](https://user-images.githubusercontent.com/680386/47966815-02e2e180-e057-11e8-93cd-087f737fbbfe.png)
**diamond bar**
![diamond bar](https://user-images.githubusercontent.com/680386/47966817-02e2e180-e057-11e8-83aa-aeb957064c38.png)
**diamond line**
![diamond line](https://user-images.githubusercontent.com/680386/47966818-037b7800-e057-11e8-9fec-c268f79882ef.png)
**diamond slider**
![diamond slider](https://user-images.githubusercontent.com/680386/47966819-037b7800-e057-11e8-9d6a-2ca9d3fbe4db.png)
**knob bar**
![knob bar](https://user-images.githubusercontent.com/680386/47966820-037b7800-e057-11e8-8d63-6dc8aa768890.png)
**knob line**
![knob line](https://user-images.githubusercontent.com/680386/47966821-037b7800-e057-11e8-8df5-c5e9a9657b59.png)
**knob slider**
![knob slider](https://user-images.githubusercontent.com/680386/47966823-037b7800-e057-11e8-9d95-0f9bdfda8012.png)

medhefgo commented:
>1. bar-bar is kinda useless for backwards seek ranges. the only thing i see, that can be done, is to do it similar to bar-line and add a margin at the bottom.

It certainly is, though alternatives look ugly. I doubt your suggestion will fare any better here.
Imho the important info is the future seekranges. In fact, mpv seems to be the only player I know that shows us past ones.

>2. a rather small problem with the knob/diamond style is that one can't see the seek ranges that are underneath the knob/diamond. for the line/slider one could overlay the range so they are visible. with the bar that is a bit harder and would look messy.

As stated earlier, this may be improved a bit my reducing the knob/diamond size in slider/line mode. I can change that if desired.

>3. for bar-line i am not a big fan that the bar for the current position is underneath the seekrage and shows through in the none seekable gaps (it changes the colour of the gaps). i personally would rather have that bar smaller.

I don't understand what you're describing here.

I have to admit that after several weeks of having mpv pick a random style every launch for testing I ended up choosing bar-line as my default. It seems to be the most useful and clear style.

Akemi commented:
1. it will at least fare better in the visibility of information. not taking visuals into account. imho not just the future seek range is an important info and a bit of consistency between the styles and the shown information would be nice too.
2. i don't think just decreasing the size is a 'solution', though yeah it's a minor problem.
3. take those screenshots below. the first one looks like we have two separate bars (bigger top, smaller below) for the current position and the seek ranges. though advancing the playback time it looks more like a small bar overlaid above a bigger bar (second screen). the current position bar fills the gaps between the seek ranges, which i find kinda irritating and distracting. it looks like the seek range bar has three different states (filled (light gray), gap (dark gray), ? (white)).

![1](https://user-images.githubusercontent.com/680386/47968338-e2238780-e068-11e8-9900-332c36f1809f.png)
![2](https://user-images.githubusercontent.com/680386/47968339-e2238780-e068-11e8-9423-7c117a906881.png)

anyway my intention of commenting was to get the discussion moving again and finally getting this merged. even if we can't get rid of these 3 rather small personal problems i won't prevent this from getting merged, because this is imo an overall improvement to the status quo.

lextra2 commented:
personally, I prefer the diamond & knob bar the most since they are, to me, the clearest way to make out the cache in the first place. also, how is having the diamond/knob hiding the cache on the small part of their location an issues, when that part would be always cached anyway?

the way I see it is, either your stream is playing, which means the part is already cached, or it is pausing, which means the stream is still downloading. in case of the former, you will see the back-cache shortly after playing some of the video, and in case of the latter you couldn't care less about a visual indicator of cache since you will have to wait for more of the stream to download anyway.

medhefgo commented:
@Akemi I gave this another go and added line-styles to the backward seek ranges when using bar seekstyle. I also added an option to separate the line seekrange from the seekbar to address your last point.

Now for the knob/diamond style I tried reducing the handle size. It does improve things a little bit for line/slider style. Though, I think the bar style should use the old/full handle size. But I also agree with @lextra2 that it's not that much of an issue (anymore) and we can finally get this merged.

Here's a preview of how things look now. I'll push the changes if you can tell me which ought to be default.
![image](https://user-images.githubusercontent.com/7892635/58084192-b7d78880-7bba-11e9-826e-c389fdae8793.png)

Akemi commented:
since this addresses my point 1 and 3, and point 2 was just a very minor complain anyway, i am completely fine with these changes. imo the 'separate' styling looks awesome from a usability standpoint. much better than the 'overlaid' one.

i would say we should make bar-bar-seperate default, since in comparison to bar-line-seperate future seek ranges are easier to see.

if you push your changes i will have another look at the code and you can have a LGTM from my side. there shouldn't be anything anymore that should block this and if anything should pop up it can be changed in a future PR.

medhefgo commented:
Pushed changes.

sfan5 commented on DOCS/man/osc.rst:
```diff
@@ -167,9 +167,8 @@ Configurable Options
-    marker with guide), or bar (fill).
-    Default pre-0.21.0 was 'slider'.
+    Sets the style of the playback position marker and overall shape
+    of the seekbar: ``bar``, ``diamond`` or ``knob``.
```
Why rename `slider` to `diamond`?

medhefgo replied:
Because both diamond and knob can be sliders now. It's part of the seekrangestyle option now.

sfan5 replied:
Right, makes sense.

sfan5 commented:
After an IRC discussion here are my personal conclusions:
* bikeshedding ensued because this PR tries to do everything at once
* round/hexagon edges are a good option to have
* diamond - slider and knob - bar/line are also good, since these weren't possible before
* in some cases the transparency for seek ranges is not strictly necessary, they could be opaque (esp. diamond - line, knob - line)
* the bar styles don't feel better than the current defaults, e.g. they look more complicated

Akemi commented on DOCS/man/osc.rst:
```diff
@@ -180,6 +179,21 @@ Configurable Options
+    Note that these will look differently based on the seekbarstyle option.
+
+``seekrangeseparate``
+    Default: true
```
yes/no, not true/false

Akemi commented on DOCS/man/osc.rst:
```diff
@@ -180,6 +179,21 @@ Configurable Options
+
+    Display seekable ranges on the seekbar. ``bar`` shows them on the full
+    height of the bar, ``line`` as a thin line and ``none`` will hide them.
+    Additionally, ``slider`` will show a permanent handle inside the seekbar
```
might mention slider won't work with bar

medhefgo commented:
Pushed documentation fixes.

Akemi commented:
we had a discussion yesterday on irc and had a few disagreements about the improvements. though we also agreed on some/most parts. that being said we would like to split these changes into a few parts if possible and take them separately.

first of all, we couldn't agree what to do with the bar style. we couldn't all see it as an improvement over the current status quo. that's something we would like to discuss a bit more and if possible take at a later point of time.

we agreed that all the knob and diamond changes are an improvement and we would like to take those first.

a few things that i noticed from the old OSC which should be fixed:
- inconsistent bar/cache styling knob vs bar/diamond (eg knob always has a bar and cache inside that bar)
- box layout with knob doesn't have a frame around the slider, in comparison to the other two options
- the box styling of the knob is slightly broken

all three of these points are addressed by your changes. so that is already a good thing.

two problems we still have with your changes:
- "the simple seek range is a white line and is inverted if a white element is at the same position" (eg the current bar and diamond styling) is gone. seems like this is a popular option and shouldn't be removed
- some seek ranges are not completely opaque but could be

to solves these problems i would like to suggest:
1. make the old styling another option (the inverted style)
2. make the seek range opacity configurable for bar and line (chose the current opacity as sensible default since otherwise it wouldn't work with all combinations)

so what should be done:
1. add the old inverted styling back as an option (seekrange=inverted), eg we would get a new knob styling out of this since the current one is already there in an improvement version by you
2. add an option to change the seek range opacity
3. leave the new bar styling out for now only keeping the old inverted options

this will leave knob and diamond with 4 distinct stylings (bar, line, slider, inverted) and bar with 1.

what should be done for the second part:
since we would only have one bar styling now (inverted) my suggestion would be analogue to the aforementioned changes, because otherwise it would be horrible inconsistent.
1. add back the inverted styling for bar, or rather keep it since we added it back before
2. add back your 4 new stylings (we didn't take in the first part)

this will leave us with 5 distinct bar stylings.

i hope this can make everyone happy. i am sorry we are asking so much of you and i personally really appreciate the work and improvements you brought with these changes, thank you.

if anything is unclear let me know and i try to clear up any uncertainties asap.

also i made [screenshots](https://0x0.st/zMpj.png) of nearly all possible old and new styling for reference.

medhefgo commented:
I am not quite sure what you want the inverted style to look like. If I get you correctly, you want all the following styles to be possible too?
![image](https://user-images.githubusercontent.com/7892635/58801141-000a9800-860a-11e9-949c-e0c9c8850bc9.png)

Akemi commented:
from your example only the last/fifth one for the bar. analogue the same for diamond and knob. for diamond see my screenshot the one right next to "bottom diamond" (the old slider). the same for knob (that styling didn't exist yet, but for consistency).

so just a centre thin white line that indicates the cache and is inverted on the position element itself.

medhefgo commented:
I added seekrangealpha and the inverted seekrangestyle option.

Akemi commented:
okay finally got some time to test again and there are still some minor things:
- i think we can't decide on a new default style right now, so could you please make the default the old default seekrangestyle=inverted
- bar-inverted should use the full size for the seekbar, that way the cache is centered. looks a bit weird right now
![bar-inverted-seekrangeseperate=no](https://user-images.githubusercontent.com/680386/59158578-3a69ae80-8abc-11e9-9ef5-430c17229a63.png)
![bar-inverted full size bar](https://user-images.githubusercontent.com/680386/59158580-3b024500-8abc-11e9-8887-5396f9f4bebd.png)
- could you make the ends of the seekranges the same form as the knob/diamond like on the new styles. it would match the frame around too.
![inverted cache endings](https://user-images.githubusercontent.com/680386/59158579-3b024500-8abc-11e9-943f-ae7e19c62cc9.png)
- with the inverted style the knob/diamond should fill the whole height like before. with the box style it gets really tiny. preferable this could be done with an option which also effects the other styles.
![fullsize diamond](https://user-images.githubusercontent.com/680386/59158581-3b024500-8abc-11e9-9f0b-81bbeb2efea8.png)
![fullsize knob](https://user-images.githubusercontent.com/680386/59158582-3b024500-8abc-11e9-83d7-4fad5717ffdf.png)

sry for all those small nitpicks. i hope those will be the last ones and we can get that merged afterwards.

medhefgo commented:
Fixed things and added a option to control the handle size. I left it at what it was previously set to for now.

Akemi commented:
bar-inverted-seekrangeseparate=no doesn't use the full height for the seek bar.

medhefgo commented:
Fixed

Akemi commented:
okay i tried to look at and test everything. LGTM now.
